### PR TITLE
feat: ads error interface

### PR DESCRIPTION
**Pull request by wseymour15**
The implementation for an error interface specifically for ads. An error can be set on the `player.ads.error()` function by passing it an object containing a `type` string and a `metadata` object containing any custom data that is desired. The bulk of the work is in [ads.js](https://github.com/videojs/videojs-contrib-ads/pull/568/files#diff-4a89f8fbc0c2d6ec50e173833d52d84816b1d7e46c024dfa1c6de39a690c7f33)

There are examples of usage in this repo, but this can also be used when contrib ads is pulled into a user's repo. Errors can be handled or set by the user.

**Proposed Changes**:
- Error interface so we can listen for ad errors in other ares of the player.
- Trigger a handful of errors
- Test Changes
- Documentation Changes

### CODE REVIEW

File: src/ads.js [hunk 1]
```diff
@@ -226,6 +231,49 @@ export default function getAds(player) {
...
+      if (err) {
+        this._error = null;
+
+        return;
+      }
+
+      this._error = err;
+
+      // TODO: log error
+      // do we want warn instead? Do we want to log here, or leave that up to when `error` is called?
```
Comment (wseymour15) [inline src/ads.js:269]: I would appreciate feedback on thoughts whether we should log here or not.

File: docs/integrator/api.md [hunk 2]
```diff
@@ -75,6 +75,38 @@ if (player.ads.inAdBreak()) {
...
+  errorType: 'ads-error-type',
+  // The key can have any name, and the value can have any type of data.
+  customData: 'custom-data'
+  ...
+}
+```
+
+To get the current ads error, the `error` function should be called with no parameters. Returns `null` if there is no error.
+```js
+const currentAd = player.ads.error();
```
Comment (dzianis-dashkevich) [inline docs/integrator/api.md:94]:
```suggestion
const currentAdError = player.ads.error();
```

File: docs/integrator/api.md [hunk 3]
```diff
@@ -75,6 +75,38 @@ if (player.ads.inAdBreak()) {
...
+```
+
+To set the current ads error, the `error` function should be called with a valid error.
+```js
+const ad = { errorType: 'ads-error-type' }
```
Comment (dzianis-dashkevich) [inline docs/integrator/api.md:99]:
```suggestion
const adError = { errorType: 'ads-error-type' }
```

File: docs/integrator/api.md [hunk 4]
```diff
@@ -75,6 +75,38 @@ if (player.ads.inAdBreak()) {
...
+
+player.ads.error(ad);
```
Comment (dzianis-dashkevich) [inline docs/integrator/api.md:101]:
```suggestion
player.ads.error(adError);
```

File: src/macros.js [hunk 5]
```diff
@@ -172,6 +172,11 @@ const replaceMacros = function(string, macros, uriEncode, overrides = {}) {
 
         string = string.replace(regex, uriEncodeIfNeeded(macros[macroName], uriEncode));
       } catch (error) {
+        player.ads.error({
+          errorType: videojs.Error.AdsMacroReplacementFailed,
+          macro: macroName
```
Comment (dzianis-dashkevich) [inline src/macros.js:177]: I guess we can pass original error here

File: src/snapshot.js [hunk 6]
```diff
@@ -196,6 +196,10 @@ export function restorePlayerSnapshot(player, callback) {
       try {
         resume();
       } catch (e) {
+        player.ads.error({
+          errorType: videojs.Error.AdsResumeContentFailed
```
Comment (dzianis-dashkevich) [inline src/snapshot.js:200]: should we pass it here as well?

**Review by dzianis-dashkevich — APPROVED** [review]

File: src/snapshot.js (hunk 6 above)
Comment (wseymour15) [reply src/snapshot.js]: Good call! I updated this in the latest commit

**Pull request merged**